### PR TITLE
Add permanent subscription option

### PR DESCRIPTION
**Pull request by JunAnnLaiIntel**
This commit
 - synchronously dump all subscription data when a subscription is modified
 - data dump is only send to the subscriber that requested the change.
 - data dump ends with a NULL management message to indicate the end of the data

### CODE REVIEW

File: clock.c [hunk 1]
```diff
@@ -311,7 +311,7 @@ static void clock_prune_subscriptions(struct clock *c)
 
 	clock_gettime(CLOCK_MONOTONIC, &now);
 	LIST_FOREACH_SAFE(s, &c->subscribers, list, tmp) {
-		if (s->expiration <= now.tv_sec) {
+		if (s->expiration != 0 && s->expiration <= now.tv_sec) {
```
Comment (yoongsiang2) [inline clock.c:316]: put a comment/note here saying The subscription will remain active if the subscription duration is set to zero.

**Comment by yoongsiang2** [comment]
remove extra space in commit msg:
image

File: clock.c [hunk 2]
```diff
@@ -311,7 +311,10 @@ static void clock_prune_subscriptions(struct clock *c)
 
 	clock_gettime(CLOCK_MONOTONIC, &now);
 	LIST_FOREACH_SAFE(s, &c->subscribers, list, tmp) {
-		if (s->expiration <= now.tv_sec) {
+		/* Subscription will remain active if the subscription
+		 * duration is set to zero
```
Comment (yoongsiang2) [inline clock.c]: add period "." at the end of the sentences.

**Comment by yoongsiang2** [comment]
maybe can improve the commit msg as below:
This patch enhances the subscription expiration logic in the clock module
by extending the functionality for subscriptions with a duration of zero.
Previously, a zero duration allowed for a single, non-recurring
subscription. With this update, a zero duration now enables a permanent
subscription, ensuring such subscriptions are not prematurely removed and
can persist indefinitely.

**Comment by yoongsiang2** [comment]
need to check it should be Christopher S Hall or Christopher S M Hall?

**Review by yoongsiang2 — APPROVED** [review]
LGTM

**Comment by AzuraTarmiziIntel** [comment]
> maybe can improve the commit msg as below: This patch enhances the subscription expiration logic in the clock module by extending the functionality for subscriptions with a duration of zero. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.

a bit too long from my point of view.  This is shorter. 
This patch enhances the subscription expiration logic in the clock module. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.

another thing, we will need to be prepared on question of why do we need to have permanent subscription .  Because in this case, zero becomes permanent subscription, so we removed the functionality of having just single-non-recurring subscription.  (which might be needed by some people) 

or do they cater for maximum value ? if maximum value is selected , it could mean permanent ? and zero would still be single-non-recurring subscription. in which we do extend functionality.

**Comment by yoongsiang2** [comment]
> > maybe can improve the commit msg as below: This patch enhances the subscription expiration logic in the clock module by extending the functionality for subscriptions with a duration of zero. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.
> 
> a bit too long from my point of view. This is shorter. This patch enhances the subscription expiration logic in the clock module. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.
> 
> another thing, we will need to be prepared on question of why do we need to have permanent subscription . Because in this case, zero becomes permanent subscription, so we removed the functionality of having just single-non-recurring subscription. (which might be needed by some people)
> 
> or do they cater for maximum value ? if maximum value is selected , it could mean permanent ? and zero would still be single-non-recurring subscription. in which we do extend functionality.

yup, we have the same concern. normally ppl use maximum value, which is -1, to represent infinite. however, the type of duration parameter is unsigned. i think we can try to use UINT16_MAX as permanent value.

**Comment by AzuraTarmiziIntel** [comment]
> > > maybe can improve the commit msg as below: This patch enhances the subscription expiration logic in the clock module by extending the functionality for subscriptions with a duration of zero. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.
> > 
> > 
> > a bit too long from my point of view. This is shorter. This patch enhances the subscription expiration logic in the clock module. Previously, a zero duration allowed for a single, non-recurring subscription. With this update, a zero duration now enables a permanent subscription, ensuring such subscriptions are not prematurely removed and can persist indefinitely.
> > another thing, we will need to be prepared on question of why do we need to have permanent subscription . Because in this case, zero becomes permanent subscription, so we removed the functionality of having just single-non-recurring subscription. (which might be needed by some people)
> > or do they cater for maximum value ? if maximum value is selected , it could mean permanent ? and zero would still be single-non-recurring subscription. in which we do extend functionality.
> 
> yup, we have the same concern. normally ppl use maximum value, which is -1, to represent infinite. however, the type of duration parameter is unsigned. i think we can try to use UINT16_MAX as permanent value.

@JunAnnLaiIntel  so what do you think ? 
I am ok, FFFF for being infinite subscription. 
0 remain as it is. 

but in this case we need to update libptpmgmt code to pass in the duration of FF instead of 0. 
will try to look it up.

File: clock.c [hunk 3]
```diff
@@ -311,7 +311,10 @@ static void clock_prune_subscriptions(struct clock *c)
 
 	clock_gettime(CLOCK_MONOTONIC, &now);
 	LIST_FOREACH_SAFE(s, &c->subscribers, list, tmp) {
-		if (s->expiration <= now.tv_sec) {
+		/* Subscription will remain active if the subscription
+		 * duration is set to zero.
```
Comment (yoongsiang2) [inline clock.c]: Subscription will not expired if duration is set to UINT16_MAX

**Review by yoongsiang2 — APPROVED** [review]
LGTM

**Pull request merged**